File: java/yass/main/ch/softappeal/yass/remote/session/Session.java
package ch.softappeal.yass.remote.session;

import ch.softappeal.yass.Closer;
import ch.softappeal.yass.Exceptions;
import ch.softappeal.yass.Nullable;
import ch.softappeal.yass.remote.Client;
import ch.softappeal.yass.remote.ExceptionReply;
import ch.softappeal.yass.remote.Reply;
import ch.softappeal.yass.remote.Request;
import ch.softappeal.yass.remote.Server;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public abstract class Session extends Client implements Closer {

    public final Connection connection;

    protected Session(final Connection connection) {
        this.connection = Objects.requireNonNull(connection);
    }

    /**
     * Called if a session has been opened.
     * Must call {@link Runnable#run()} (possibly in an own thread).
     */
    protected abstract void dispatchOpened(Runnable runnable) throws Exception;

    /**
     * Called for an incoming request.
     * Must call {@link Runnable#run()} (possibly in an own thread).
     */
    protected abstract void dispatchServerInvoke(Server.Invocation invocation, Runnable runnable) throws Exception;

    private Server server;
    /**
     * Gets the server of this session. Called only once after creation of session.
     * This implementation returns {@link Server#EMPTY}.
     */
    protected Server server() throws Exception {
        return Server.EMPTY;
    }

    /**
     * Called when the session has been opened.
     * This implementation does nothing.
     * Due to race conditions or exceptions it could not be called or be called after {@link #closed(Exception)}.
     * @see SessionFactory#create(Connection)
     */
    protected void opened() throws Exception {
        // empty
    }

    /**
     * Called once when the session has been closed.
     * This implementation does nothing.
     * @param exception if (exception == null) regular close else reason for close
     * @see SessionFactory#create(Connection)
     */
    protected void closed(final @Nullable Exception exception) throws Exception {
        // empty
    }

    private final AtomicBoolean closed = new AtomicBoolean(true);
    public final boolean isClosed() {
        return closed.get();
    }

    private void unblockPromises() {
        for (final var invocation : new ArrayList<>(requestNumber2invocation.values())) {
            try {
                invocation.settle(new ExceptionReply(new SessionClosedException()));
            } catch (final Exception ignore) {
                // empty
            }
        }
    }

    private void close(final boolean sendEnd, final @Nullable Exception exception) {
        if (closed.getAndSet(true)) {
            return;
        }
        try {
            try {
                try {
                    unblockPromises();
                } finally {
                    closed(exception);
                }
                if (sendEnd) {
                    connection.write(Packet.END);
                }
            } finally {
                connection.closed();
            }
        } catch (final Exception e) {
            throw Exceptions.wrap(e);
        }
    }

    /**
     * Must be called if communication has failed.
     * This method is idempotent.
     */
    public static void close(final Session session, final Exception e) {
        session.close(false, Objects.requireNonNull(e));
    }

    private void closeThrow(final Exception e) throws Exception {
        try {
            close(this, e);
        } catch (final Exception e2) {
            e.addSuppressed(e2);
        }
        throw e;
    }

    /**
     * @see #close(Session, Exception)
     */
    public static void closeThrow(final Session session, final Exception e) throws Exception {
        session.closeThrow(e);
    }

    @Override public void close() {
        close(true, null);
    }

    private void serverInvoke(final int requestNumber, final Request request) throws Exception {
        final var invocation = server.invocation(true, request);
        dispatchServerInvoke(invocation, () -> {
            try {
                invocation.invoke(reply -> {
                    if (!invocation.methodMapping.oneWay) {
                        try {
                            connection.write(new Packet(requestNumber, reply));
                        } catch (final Exception e) {
                            closeThrow(e);
                        }
                    }
                });
            } catch (final Exception e) {
                close(this, e);
            }
        });
    }

    /**
     * note: it's not worth to use {@link ConcurrentHashMap} here
     */
    private final Map<Integer, Invocation> requestNumber2invocation = Collections.synchronizedMap(new HashMap<>(16));

    private void received(final Packet packet) throws Exception {
        try {
            if (packet.isEnd()) {
                close(false, null);
                return;
            }
            final var message = packet.message();
            if (message instanceof Request) {
                serverInvoke(packet.requestNumber(), (Request)message);
            } else {
                requestNumber2invocation.remove(packet.requestNumber()).settle((Reply)message); // client invoke
            }
        } catch (final Exception e) {
            closeThrow(e);
        }
    }

    /**
     * Must be called if a packet has been received.
     * It must also be called if {@link Packet#isEnd()}; however, it must not be called again after that.
     */
    public static void received(final Session session, final Packet packet) throws Exception {
        session.received(packet);
    }

    private final AtomicInteger nextRequestNumber = new AtomicInteger(Packet.END_REQUEST_NUMBER);

    @Override protected final void invoke(final Client.Invocation invocation) throws Exception {
        if (isClosed()) {
            throw new SessionClosedException();
        }
        invocation.invoke(true, request -> {
            try {
                int requestNumber;
                do { // we can't use END_REQUEST_NUMBER as regular requestNumber
                    requestNumber = nextRequestNumber.incrementAndGet();
                } while (requestNumber == Packet.END_REQUEST_NUMBER);
                if (!invocation.methodMapping.oneWay) {
                    requestNumber2invocation.put(requestNumber, invocation);
                    if (isClosed()) {
                        unblockPromises(); // needed due to race conditions
                    }
                }
                connection.write(new Packet(requestNumber, request));
            } catch (final Exception e) {
                closeThrow(e);
            }
        });
    }

    private void created() throws Exception {
        closed.set(false);
        try {
            server = Objects.requireNonNull(server());
            dispatchOpened(() -> {
                try {
                    opened();
                } catch (final Exception e) {
                    close(this, e);
                }
            });
        } catch (final Exception e) {
            closeThrow(e);
        }
    }

    public static Session create(final SessionFactory sessionFactory, final Connection connection) throws Exception {
        final var session = sessionFactory.create(connection);
        session.created();
        return session;
    }

}

File: kotlin/yass/main/ch/softappeal/yass/remote/session/Session.kt
package ch.softappeal.yass.remote.session

import ch.softappeal.yass.*
import ch.softappeal.yass.remote.*
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.atomic.*

interface Connection {
    /** Called if a packet has to be written out. */
    @Throws(Exception::class)
    fun write(packet: Packet)

    /** Called once if the connection has been closed. */
    @Throws(Exception::class)
    fun closed()
}

class SessionClosedException : RuntimeException()

abstract class Session : Client(), AutoCloseable {
    private lateinit var _connection: Connection
    val connection: Connection get() = _connection

    private lateinit var server: Server

    private val closed = AtomicBoolean(true)
    val isClosed get() = closed.get()

    /** note: it's not worth to use [ConcurrentHashMap] here */
    private val requestNumber2invocation = Collections.synchronizedMap(HashMap<Int, ClientInvocation>(16))

    private val nextRequestNumber = AtomicInteger(EndRequestNumber)

    internal fun iCreated(connection: Connection) {
        server = requireNotNull(server())
        _connection = connection
        closed.set(false)
        dispatchOpened {
            try {
                opened()
            } catch (e: Exception) {
                close(e)
            }
        }
    }

    /** Called if a session has been opened. Must call [action] (possibly in an own thread). */
    @Throws(Exception::class)
    protected abstract fun dispatchOpened(action: () -> Unit)

    /** Called for an incoming request. Must call [action] (possibly in an own thread). */
    @Throws(Exception::class)
    protected abstract fun dispatchService(invocation: ServiceInvocation, action: () -> Unit)

    /** Gets the server of this session. Called only once after creation of session. */
    @Throws(Exception::class)
    protected open fun server(): Server = EmptyServer

    @Throws(Exception::class)
    protected open fun opened() {
        // empty
    }

    /** If ([exception] == null) regular close else reason for close. */
    @Throws(Exception::class)
    protected open fun closed(exception: Exception?) {
        // empty
    }

    override fun close() =
        iClose(true, null)

    private fun settlePendingInvocations() {
        for (invocation in requestNumber2invocation.values.toList())
            try {
                invocation.settle(ExceptionReply(SessionClosedException()))
            } catch (ignore: Exception) {
            }
    }

    internal fun iClose(sendEnd: Boolean, exception: Exception?) {
        if (closed.getAndSet(true)) return
        try {
            try {
                settlePendingInvocations()
            } finally {
                closed(exception)
            }
            if (sendEnd) _connection.write(EndPacket)
        } finally {
            _connection.closed()
        }
    }

    private fun closeThrow(e: Exception) {
        addSuppressed(e) { close(e) }
        throw e
    }

    private fun serviceInvoke(requestNumber: Int, request: Request) {
        val invocation = server.invocation(true, request)
        dispatchService(invocation) {
            try {
                invocation.invoke { reply ->
                    if (!invocation.methodMapping.oneWay) {
                        try {
                            _connection.write(Packet(requestNumber, reply))
                        } catch (e: Exception) {
                            closeThrow(e)
                        }
                    }
                }
            } catch (e: Exception) {
                closeThrow(e)
            }
        }
    }

    internal fun iReceived(packet: Packet) {
        try {
            if (packet.isEnd) {
                iClose(false, null)
                return
            }
            when (val message = packet.message) {
                is Request -> serviceInvoke(packet.requestNumber, message)
                is Reply -> requestNumber2invocation.remove(packet.requestNumber)!!.settle(message)
            }
        } catch (e: Exception) {
            closeThrow(e)
        }
    }

    final override fun invoke(invocation: ClientInvocation) {
        if (isClosed) throw SessionClosedException()
        try {
            invocation.invoke(true) { request ->
                try {
                    var requestNumber: Int
                    do { // we can't use END_REQUEST_NUMBER as regular requestNumber
                        requestNumber = nextRequestNumber.incrementAndGet()
                    } while (requestNumber == EndRequestNumber)
                    if (!invocation.methodMapping.oneWay) {
                        requestNumber2invocation[requestNumber] = invocation
                        if (isClosed) settlePendingInvocations() // needed due to race conditions
                    }
                    _connection.write(Packet(requestNumber, request))
                } catch (e: Exception) {
                    closeThrow(e)
                }
            }
        } catch (e: Exception) {
            closeThrow(e)
        }
    }
}

/** Must be called if communication has failed. This method is idempotent. */
fun Session.close(e: Exception) = iClose(false, e)

/**
 * Must be called if a packet has been received.
 * It must also be called if [Packet.isEnd]; however, it must not be called again after that.
 */
fun Session.received(packet: Packet) = iReceived(packet)

fun Session.created(connection: Connection) = iCreated(connection)

typealias SessionFactory = () -> Session

open class SimpleSession(protected val dispatchExecutor: Executor) : Session() {
    override fun dispatchOpened(action: () -> Unit) = dispatchExecutor.execute { action() }
    override fun dispatchService(invocation: ServiceInvocation, action: () -> Unit) =
        dispatchExecutor.execute { action() }
}
